### DISCUSSION TITLE
Laser/highlighter design notes

### DISCUSSION THREAD

**Discussion by mjg** [xournalpp/xournalpp#3043] [Ideas]
We have a few suggestions around implementing a laser or highlighter feature in xournalpp (#1778, #1822, #3023), and there are external tools like [highlight-pointer](https://github.com/swillner/highlight-pointer) or [Projecteur](https://github.com/jahnf/Projecteur). Playing around with them I noticed a few fundamental decisions to make, which this discussion is meant to help.
### Use cases
Virtual laser or highlighting is used for live presentation or recording. There are several technical scenarios:

- present via beamer or similar (direct clone of your own screen)
- record or screen share the xournalpp window
- record or screen share the screen on which xournalpp display its window

Depending on the system and software you use for recording/screen sharing) it is able to capture an application cursor. This poses the first technical restriction if we want to support all use cases in all environments.

### Implementation approaches
There are three ways in which xournalpp or an external app can draw a laser highlight:

1. change the actual cursor (e.g. current xounalpp position highlighter)
2. draw the highlight as a window overlay (highlight-pointer: X11, Projecteur: compositor)
3. draw inside the application window

Method 1 works only if your recording/sharing environment is able to capture the application cursor.
Method 2 works only if you record/share a whole screen, which you might not want so that you avoid accidental pop-up notifications and such (but also an overlay window). In addition, X11 works only in X11, compositing works only in compositing environments, obviously.
Method 3 seems to be the only approach that could work in all setups and environments - if you can record/share what xournalpp draws it will contain anything it draws, obviously. The only issue could be timing here.

### Design approaches
Various ways of highlighting have been suggested or are in use. Time-wise:
- cursor-like (follows or is cursor, e.g. the current xournalpp position highlighter)
- cursor-like with an evaporating tail
- highlighter-like, i.e. like the existing highlighter-tool with either a fade-off after tool-release or button-release

Space-wise:
- opaque disc or other shape
- semi-transparent disc or other shape
- transparent disc or rather shape with semi-transparent outside

Projecteur has a nice way of unifying all three within a simple set of settings.

**Comment by fsbox666** [comment]
I was hesitating to start a new thread (with less IT expertise... sorry), since I wanted suggest a probably relatively simple useful implementation of the design approach suggested here "cursor-like with evaporating tail", except that I suggest a "tail" that remains visible until the pen is lifted off the tablet [as in GoodNotes](https://support.goodnotes.com/hc/en-us/articles/7451632333327-Focus-Attention-with-Laser-Pointer-): this is amazingly useful, especially for teaching(*), but also beyond that. 

Thank you for providing xournallpp, it is truly outstanding.

_(*)This would add to xournal++'s (already very good) configurable pointer. I use xournalpp mainly, but not only, for teaching anatomy and, thus, drawing in the auditorium. Before, I have had the opportunity to use GoodNotes on and iPad and that laser pointer is just great to temporarily "draw" things, show certain areas, ... and, as I have seen with colleagues in diverse other domains, focus attention on a specific area, text, formula, drawing... etc). I don't code, so I don't realise how much of an effort the implementation of this kind of a tool would be, obviously..._
